x-pack/plugins/event_log/server/es/cluster_client_adapter.ts

```diff
@@ -238,41 +239,112 @@ export class ClusterClientAdapter<TDoc extends { body: AliasAny; index: string }
       });
       throw err;
     }
+    const savedObjectsQueryMust: estypes.QueryDslQueryContainer[] = [
+      {
+        term: {
+          'kibana.saved_objects.rel': {
+            value: SAVED_OBJECT_REL_PRIMARY,
+          },
+        },
+      },
+      {
+        term: {
+          'kibana.saved_objects.type': {
+            value: type,
+          },
+        },
+      },
+      // @ts-expect-error undefined is not assignable as QueryDslTermQuery value
+      namespaceQuery,
+    ];
+    if (!legacyIds) {
+      savedObjectsQueryMust.push({
+        terms: {
+          // default maximum of 65,536 terms, configurable by index.max_terms_count
+          'kibana.saved_objects.id': ids,
+        },
+      });
+    }
     const musts: estypes.QueryDslQueryContainer[] = [
       {
         nested: {
           path: 'kibana.saved_objects',
           query: {
             bool: {
-              must: [
-                {
-                  term: {
-                    'kibana.saved_objects.rel': {
-                      value: SAVED_OBJECT_REL_PRIMARY,
+              must: reject(savedObjectsQueryMust, isUndefined),
+            },
+          },
+        },
+      },
+    ];
+    if (legacyIds) {
```

Comment:
**nit:** not sure if needed, but in case Elasticsearch doesn't like empty arrays.

```suggestion
    if (legacyIds && legacyIds.length > 0) {
```